README
NAME
    Catalyst::View::Email - Send Email from Catalyst

SYNOPSIS
    This module sends out emails from a stash key specified in the
    configuration settings.

CONFIGURATION
    WARNING: since version 0.10 the configuration options slightly changed!

    Use the helper to create your View:

        $ script/myapp_create.pl view Email Email

    In your app configuration:

        __PACKAGE__->config(
            'View::Email' => {
                # Where to look in the stash for the email information.
                # 'email' is the default, so you don't have to specify it.
                stash_key => 'email',
                # Define the defaults for the mail
                default => {
                    # Defines the default content type (mime type). Mandatory
                    content_type => 'text/plain',
                    # Defines the default charset for every MIME part with the 
                    # content type text.
                    # According to RFC2049 a MIME part without a charset should
                    # be treated as US-ASCII by the mail client.
                    # If the charset is not set it won't be set for all MIME parts
                    # without an overridden one.
                    # Default: none
                    charset => 'utf-8'
                },
                # Setup how to send the email
                # all those options are passed directly to Email::Sender::Simple
                sender => {
                    # if mailer doesn't start with Email::Sender::Simple::Transport::,
                    # then this is prepended.
                    mailer => 'SMTP',
                    # mailer_args is passed directly into Email::Sender::Simple 
                    mailer_args => {
                        host     => 'smtp.example.com', # defaults to localhost
                        username => 'username',
                        password => 'password',
                }
              }
            }
        );

NOTE ON SMTP
    If you use SMTP and don't specify host, it will default to localhost and
    attempt delivery. This often means an email will sit in a queue and not
    be delivered.

SENDING EMAIL
    Sending email is just filling the stash and forwarding to the view:

        sub controller : Private {
            my ( $self, $c ) = @_;

            $c->stash->{email} = {
                to      => 'jshirley@gmail.com',
                cc      => 'abraxxa@cpan.org',
                from    => 'no-reply@foobar.com',
                subject => 'I am a Catalyst generated email',
                body    => 'Body Body Body',
            };
        
            $c->forward( $c->view('Email') );
        }

    Alternatively you can use a more raw interface and specify the headers
    as an array reference like it is passed to Email::MIME::Creator. Note
    that you may also mix both syntaxes if you like ours better but need to
    specify additional header attributes. The attributes are appended to the
    header array reference without overwriting contained ones.

        $c->stash->{email} = {
            header => [
                To      => 'jshirley@gmail.com',
                Cc      => 'abraxxa@cpan.org',
                From    => 'no-reply@foobar.com',
                Subject => 'Note the capitalization differences',
            ],
            body => qq{Ain't got no body, and nobody cares.},
            # Or, send parts
            parts => [
                Email::MIME->create(
                    attributes => {
                        content_type => 'text/plain',
                        disposition  => 'attachment',
                        charset      => 'US-ASCII',
                    },
                    body => qq{Got a body, but didn't get ahead.},
                )
            ],
        };

HANDLING ERRORS
    If the email fails to send, the view will die (throw an exception).
    After your forward to the view, it is a good idea to check for errors:

        $c->forward( $c->view('Email') );
    
        if ( scalar( @{ $c->error } ) ) {
            $c->error(0); # Reset the error condition if you need to
            $c->response->body('Oh noes!');
        } else {
            $c->response->body('Email sent A-OK! (At least as far as we can tell)');
        }

USING TEMPLATES FOR EMAIL
    Now, it's no fun to just send out email using plain strings. Take a look
    at Catalyst::View::Email::Template to see how you can use your favourite
    template engine to render the mail body.

METHODS
    new Validates the base config and creates the Email::Sender::Simple
        object for later use by process.

    process($c)
        The process method does the actual processing when the view is
        dispatched to.

        This method sets up the email parts and hands off to
        Email::Sender::Simple to handle the actual email delivery.

    setup_attributes($c, $attr)
        Merge attributes with the configured defaults. You can override this
        method to return a structure to pass into generate_message which
        subsequently passes the return value of this method to
        Email::MIME->create under the "attributes" key.

    generate_message($c, $attr)
        Generate a message part, which should be an Email::MIME object and
        return it.

        Takes the attributes, merges with the defaults as necessary and
        returns a message object.

TROUBLESHOOTING
    As with most things computer related, things break. Email even more so.
    Typically any errors are going to come from using SMTP as your sending
    method, which means that if you are having trouble the first place to
    look is at Email::Sender::Transport::SMTP. This module is just a wrapper
    for Email::Sender::Simple, so if you get an error on sending, it is
    likely from there anyway.

    If you are using SMTP and have troubles sending, whether it is
    authentication or a very bland "Can't send" message, make sure that you
    have Net::SMTP and, if applicable, Net::SMTP::SSL installed.

    It is very simple to check that you can connect via Net::SMTP, and if
    you do have sending errors the first thing to do is to write a simple
    script that attempts to connect. If it works, it is probably something
    in your configuration so double check there. If it doesn't, well, keep
    modifying the script and/or your mail server configuration until it
    does!

SEE ALSO
  Catalyst::View::Email::Template - Send fancy template emails with Cat
  Catalyst::Manual - The Catalyst Manual
  Catalyst::Manual::Cookbook - The Catalyst Cookbook
AUTHORS
    J. Shirley <jshirley@gmail.com>

    Alexander Hartmaier <abraxxa@cpan.org>

CONTRIBUTORS
    (Thanks!)

    Matt S Trout

    Daniel Westermann-Clark

    Simon Elliott <cpan@browsing.co.uk>

    Roman Filippov

    Lance Brown <lance@bearcircle.net>

    Devin Austin <dhoss@cpan.org>

    Chris Nehren <apeiron@cpan.org>

COPYRIGHT
    Copyright (c) 2007 - 2009 the Catalyst::View::Email "AUTHORS" and
    "CONTRIBUTORS" as listed above.

LICENSE
    This library is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Catalyst::View::Email.

0.36    xxx
        - call `use_ok` once per module https://github.com/dhoss/catalyst-view-email/commit/432d95d0d52359cd4c7b6b862758c6801ca664dc

0.34    2014-05-17
        - switched from Class::MOP::load_class to Module::Runtime::require_module per
          cf. https://rt.cpan.org/Public/Bug/Display.html?id=91013
        - canonical repo now on github as noted in Makefile.PL
        - typo fix per https://rt.cpan.org/Public/Bug/Display.html?id=87370
        - packaged with /usr/bin/gnutar on OSX
          cf. https://rt.cpan.org/Public/Bug/Display.html?id=83970

0.33    2013-02-13
        - applied patch to fix Moo->Moose inflation that Email::Sender does per https://rt.cpan.org/Ticket/Display.html?id=83241
        - removed inc/ so running perl Makefile.PL would do so in author mode
        - removed META.yml, MANIFEST, and various other files that shouldn't be in git
        - updated Makefile.PL to use version_from "lib/Catalyst/View/Email.pm" per https://rt.cpan.org/Ticket/Display.html?id=83195&results=0eb39711cb23f5f318a3738cf7d089df

0.32    2013-01-15
        - Removed Bcc references per https://rt.cpan.org/Public/Bug/Display.html?id=75775
        - Fixed documentation bug, changed 'Host' to 'host' per https://rt.cpan.org/Ticket/Display.html?id=58748
        - Added patch (with additional tests) to enable proper content_type setting for email templates
          per https://rt.cpan.org/Ticket/Display.html?id=66495&results=73b48805c7d356914e91ea95e31ddef6
        - Removed unneccesary test
        - Added documentation for specifying envelope info per https://rt.cpan.org/Ticket/Display.html?id=64162

0.31    2011-01-17
        - Added handling and test for encoding and quoting parts RT #56391 

0.30    2010-05-22
        - Added Test::Requires to 06config.t

0.29    2010-05-21
        - Using Test::Requires for Mason and Template::Toolkit dependency check
        - Removed Template::Toolkit and Mason as prereqs

0.28    2010-05-14
        - Added  version numbers to Helpers

0.27    2010-03-23
        - made ::View::TT and ::View::Mason hard deps
        - adjusted tests to reflect the aforementioned
        - patches welcome

0.25_02 2010-02-21
        - cleaned up tests
        - made Test::More 0.88 a prereq

0.25_01 2010-02-21
        - updated/corrected docs
        - added proper svn repo url

0.25    2010-02-01
        - converted to Module::Install

0.23    2010-01-30
        - removed Email::Send::Test from t/lib/* views

0.22    2010-01-28
        - small fixes, version bumpage etc

0.21_01 2010-01-27
        - dev release to test recent fixes

0.21    2010-01-27
        - provided a sane MANIFEST.SKIP for dzil

0.20    2010-01-26
        - fix transport instantiation thanks to Chris Nehren <apeiron@cpan.org>

0.19    2010-01-18
        - fixed optional dependencies
		- added META stuff back in

0.18    2010-01-17
        - Fixed version dependencies for failing regressions
		- added Email::Send::Test as a dependency

0.17    2010-01-14
        - fixed MANIFEST

0.16    2010-01-12
        - Fixed a Makefile.PL issue

0.15    2010-01-12
        - Fixed teeny version issue with Email::Sender::Simple

0.14    2010-01-09
        - Ported over to using Email::Sender::Simple, and Moose
		- Fixed Mason test so that it doesn't choke, as around 'process' seems to make
		  it die.

0.13    2009-02-22 14:06:00
        - bcc: POD corrections thanks to Lance Brown <lance@bearcircle.net>

0.12    2009-01-22 06:52:00
        - Fixing tests for new versions of MIME::Creator
        - Better structure of the code so that ::Template can also handle
          plain text views
        - Added onto troubleshooting

0.11    2008-07-04 09:14:00
        - Fixing a bug where content-type was ignored so multipart/alternative
          failed.  RT #32215

0.10    2007-11-22 23:00:00
        - Refactored by Alexander Hartmaier with api changes
          and POD improvements

0.06
        - Fixing some slight issues with configuration not being handled
          appropriately (thanks dwc and mst)
0.05
        - Better support for configuration
        - Adding Mason tests

0.04
        - Fixing MANIFEST, distribution

0.03
        - Improved handling of configuration, thanks to mst.

0.02
        - Allow configuration from application or ConfigLoader
        - Minor POD updates

0.01  2007-06-03
        - Initial release


t/lib/TestApp/Controller/Root.pm
package  # Hide from PAUSE
    TestApp::Controller::Root;

use base qw(Catalyst::Controller);

use Encode;

sub default : Private {
    my ( $self, $c ) = @_;

    $c->res->body(qq{Nothing Here});
}

sub email : Global('email') {
    my ($self, $c, @args) = @_;

    my $time = $c->req->params->{time} || time;

    $c->stash->{email} = {
        to      => 'test-email@example.com',
        from    => 'no-reply@example.com',
        subject => 'Email Test',
        body    => "Email Sent at: $time"
    };

    $c->forward('TestApp::View::Email');

    if ( scalar( @{ $c->error } ) ) {
        $c->res->status(500);
        $c->res->body('Email Failed');
    } else {
        $c->res->body('Plain Email Ok');
    }
}

sub email_app_config : Global('email_app_config') {
    my ($self, $c, @args) = @_;

    my $time = $c->req->params->{time} || time;

    $c->stash->{email} = {
        to      => 'test-email@example.com',
        from    => 'no-reply@example.com',
        subject => 'Email Test',
        body    => "Email Sent at: $time"
    };

    $c->forward('TestApp::View::Email::AppConfig');

    if ( scalar( @{ $c->error } ) ) {
        $c->res->status(500);
        $c->res->body('Email Failed');
    } else {
        $c->res->body('Plain Email Ok');
    }
}

sub template_email : Global('template_email') {
    my ($self, $c, @args) = @_;

    $c->stash->{time} = $c->req->params->{time} || time;

    $c->stash->{email} = {
        to           => 'test-email@example.com',
        from         => 'no-reply@example.com',
        subject      => 'Just a test',
        content_type => 'multipart/alternative',
        templates => [
            {
                template        => 'text_plain/test.tt',
                content_type    => 'text/plain',
            },
            {
                view            => 'TT',
                template        => 'text_html/test.tt',
                content_type    => 'text/html',
            },
        ],
    };

    $c->forward('TestApp::View::Email::Template');    

    if ( scalar( @{ $c->error } ) ) {
        $c->res->status(500);
        $c->res->body('Template Email Failed');
    } else {
        $c->res->body('Template Email Ok');
    }
}

sub template_email_utf8 : Global('template_email_utf8') {
    my ($self, $c, @args) = @_;

    $c->stash->{time} = $c->req->params->{time} || time;

    $c->stash->{chars} = decode('utf-8', "✔ ✈ ✉");

    $c->stash->{email} = {
        to           => 'test-email@example.com',
        from         => 'no-reply@example.com',
        subject      => 'Just a test',
        content_type => 'multipart/alternative',
       templates => [
            {
                template        => 'text_plain/test.tt',
                content_type    => 'text/plain',
                charset      => 'utf-8',
                encoding     => 'quoted-printable',
            },
            {
                view            => 'TT',
                template        => 'text_html/test_utf8.tt',
                content_type    => 'text/html',
                charset      => 'utf-8',
                encoding     => 'quoted-printable',
             },
        ],
    };

    $c->forward('TestApp::View::Email::Template');    

    if ( scalar( @{ $c->error } ) ) {
        $c->res->status(500);
        $c->res->body('Template Email Failed');
    } else {
        $c->res->body('Template Email Ok');
    }
}



sub template_email_app_config : Global('template_email_app_config') {
    my ($self, $c, @args) = @_;

    $c->stash->{time} = $c->req->params->{time} || time;

    $c->stash->{template_email} = {
        to      => 'test-email@example.com',
        from    => 'no-reply@example.com',
        subject => 'Just a test',
        templates => [
            {
                template        => 'text_plain/test.tt',
                content_type    => 'text/plain',
            },
            {
                view            => 'TT',
                template        => 'text_html/test.tt',
                content_type    => 'text/html',
            },
        ],
    };

    $c->forward('TestApp::View::Email::Template::AppConfig');

    if ( scalar( @{ $c->error } ) ) {
        $c->res->status(500);
        $c->res->body('Template Email Failed');
    } else {
        $c->res->body('Template Email Ok');
    }
}

sub mason_email : Global('mason_email') {
    my ($self, $c, @args) = @_;

    $c->stash->{time} = $c->req->params->{time} || time;

    $c->stash->{email} = {
        to      => 'test-email@example.com',
        from    => 'no-reply@example.com',
        subject => 'Just a test',
        templates => [
            {
                view            => 'Mason',
		        template        => 'text_plain/test.m',
                content_type    => 'text/plain',
            },
            {
                view            => 'Mason',
                template        => 'text_html/test.m',
                content_type    => 'text/html',
            },
        ],
    };

    $c->forward('TestApp::View::Email::Template');    

    if ( scalar( @{ $c->error } ) ) {
        $c->res->status(500);
        $c->res->body('Mason Email Failed');
    } else {
        $c->res->body('Mason Email Ok');
    }
}


1;


Makefile.PL
use inc::Module::Install 0.91;
test_requires 'Test::Requires';

name 'Catalyst-View-Email';
all_from 'lib/Catalyst/View/Email.pm';
author 'J. Shirley <jshirley@gmail.com>';
license 'perl';
version_from 'lib/Catalyst/View/Email.pm';
test_requires 'Test::More';

requires 'MIME::Base64'          => '3.08';
requires 'Authen::SASL'          => '2.13';
requires 'Catalyst'              => '5.7';
requires 'Moose'                 => '0.93';
requires 'parent'                => '0.223';
requires 'Email::Sender::Simple' => '0.100110';
requires 'Email::MIME'           => '1.859';
requires 'Email::MIME::Creator'  => '1.455';
requires 'Test::More'            => '0.88';
requires 'Module::Runtime'       => '0.014';
#requires 'Catalyst::View::TT'    => '0.31';
#requires 'Catalyst::View::Mason' => '0.18';

# Rebuild README for maintainers
if ($Module::Install::AUTHOR) {
    system("pod2text lib/Catalyst/View/Email.pm > README");
}

realclean_files 'README';

resources 'IRC'     => 'irc://irc.perl.org/#catalyst';
resources 'license' => 'http://dev.perl.org/licenses/';
resources 'repository' =>
  'http://github.com/dhoss/catalyst-view-email';
resources 'MailingList' =>
  'http://lists.scsys.co.uk/cgi-bin/mailman/listinfo/catalyst';
resources 'bugtracker' => 
  'https://github.com/dhoss/catalyst-view-email/issues';

auto_provides;
auto_install;
WriteAll;


lib/Catalyst/View/Email.pm
package Catalyst::View::Email;

use Moose;
use Carp;

use Encode qw(encode decode);
use Email::Sender::Simple qw/ sendmail /;
use Email::MIME::Creator;
use Module::Runtime;
extends 'Catalyst::View';

our $VERSION = '0.36';
$VERSION = eval $VERSION;

has 'mailer' => (
    is      => 'rw',
    isa     => 'Str',
    lazy    => 1,
    default => sub { "sendmail" }
);

has '_mailer_obj' => (
    is      => 'rw',
    does    => 'Email::Sender::Transport',
    lazy    => 1,
    builder => '_build_mailer_obj',
);

has 'stash_key' => (
    is      => 'rw',
    isa     => 'Str',
    lazy    => 1,
    default => sub { "email" }
);

has 'default' => (
    is      => 'rw',
    isa     => 'HashRef',
    default => sub { { content_type => 'text/plain' } },
    lazy    => 1,
);

has 'sender' => (
    is      => 'rw',
    isa     => 'HashRef',
    lazy    => 1,
    default => sub { { mailer => shift->mailer } }
);

has 'content_type' => (
    is      => 'rw',
    isa     => 'Str',
    default => sub { shift->default->{content_type} },
    lazy    => 1,
);

=head1 NAME

Catalyst::View::Email - Send Email from Catalyst

=head1 SYNOPSIS

This module sends out emails from a stash key specified in the
configuration settings.

=head1 CONFIGURATION

WARNING: since version 0.10 the configuration options slightly changed!

Use the helper to create your View:
    
    $ script/myapp_create.pl view Email Email

In your app configuration:

    __PACKAGE__->config(
        'View::Email' => {
            # Where to look in the stash for the email information.
            # 'email' is the default, so you don't have to specify it.
            stash_key => 'email',
            # Define the defaults for the mail
            default => {
                # Defines the default content type (mime type). Mandatory
                content_type => 'text/plain',
                # Defines the default charset for every MIME part with the 
                # content type text.
                # According to RFC2049 a MIME part without a charset should
                # be treated as US-ASCII by the mail client.
                # If the charset is not set it won't be set for all MIME parts
                # without an overridden one.
                # Default: none
                charset => 'utf-8'
            },
            # Setup how to send the email
            # all those options are passed directly to Email::Sender::Simple
            sender => {
                # if mailer doesn't start with Email::Sender::Simple::Transport::,
                # then this is prepended.
                mailer => 'SMTP',
                # mailer_args is passed directly into Email::Sender::Simple 
                mailer_args => {
                    host     => 'smtp.example.com', # defaults to localhost
                    sasl_username => 'sasl_username',
                    sasl_password => 'sasl_password',
            }
          }
        }
    );

=head1 NOTE ON SMTP

If you use SMTP and don't specify host, it will default to localhost and
attempt delivery. This often means an email will sit in a queue and
not be delivered.

=cut

=head1 SENDING EMAIL

Sending email is just filling the stash and forwarding to the view:

    sub controller : Private {
        my ( $self, $c ) = @_;

        $c->stash->{email} = {
            to      => 'jshirley@gmail.com',
            cc      => 'abraxxa@cpan.org',
            from    => 'no-reply@foobar.com',
            subject => 'I am a Catalyst generated email',
            body    => 'Body Body Body',
        };
        
        $c->forward( $c->view('Email') );
    }

Alternatively you can use a more raw interface and specify the headers as
an array reference like it is passed to L<Email::MIME::Creator>.
Note that you may also mix both syntaxes if you like ours better but need to
specify additional header attributes.
The attributes are appended to the header array reference without overwriting
contained ones.

    $c->stash->{email} = {
        header => [
            To      => 'jshirley@gmail.com',
            Cc      => 'abraxxa@cpan.org',
            Bcc     => join ',', qw/hidden@secret.com hidden2@foobar.com/,
            From    => 'no-reply@foobar.com',
            Subject => 'Note the capitalization differences',
        ],
        body => qq{Ain't got no body, and nobody cares.},
        # Or, send parts
        parts => [
            Email::MIME->create(
                attributes => {
                    content_type => 'text/plain',
                    disposition  => 'attachment',
                    charset      => 'US-ASCII',
                },
                body => qq{Got a body, but didn't get ahead.},
            )
        ],
    };

You can set the envelope sender and recipient as well:

  $c->stash->{email} = {

    envelope_from => 'envelope-from@example.com',
    from          => 'header-from@example.com',

    envelope_to   => [ 'foo@example.com', 'bar@example.com' ],
    to            => 'Undisclosed Recipients:;',

    ...
  };

=head1 HANDLING ERRORS

If the email fails to send, the view will die (throw an exception).
After your forward to the view, it is a good idea to check for errors:
    
    $c->forward( $c->view('Email') );
    
    if ( scalar( @{ $c->error } ) ) {
        $c->error(0); # Reset the error condition if you need to
        $c->response->body('Oh noes!');
    } else {
        $c->response->body('Email sent A-OK! (At least as far as we can tell)');
    }

=head1 USING TEMPLATES FOR EMAIL

Now, it's no fun to just send out email using plain strings.
Take a look at L<Catalyst::View::Email::Template> to see how you can use your
favourite template engine to render the mail body.

=head1 METHODS

=over 4

=item new

Validates the base config and creates the L<Email::Sender::Simple> object for later use
by process.

=cut

sub BUILD {
    my $self = shift;

    my $stash_key = $self->stash_key;
    croak "$self stash_key isn't defined!"
      if ( $stash_key eq '' );

}

sub _build_mailer_obj {
    my ($self) = @_;
    my $transport_class = ucfirst $self->sender->{mailer};

    # borrowed from Email::Sender::Simple -- apeiron, 2010-01-26
    if ( $transport_class !~ /^Email::Sender::Transport::/ ) {
        $transport_class = "Email::Sender::Transport::$transport_class";
    }

    Module::Runtime::require_module($transport_class);

    return $transport_class->new( $self->sender->{mailer_args} || {} );
}

=item process($c)

The process method does the actual processing when the view is dispatched to.

This method sets up the email parts and hands off to L<Email::Sender::Simple> to handle
the actual email delivery.

=cut

sub process {
    my ( $self, $c ) = @_;

    croak "Unable to send mail, bad mail configuration"
      unless $self->sender->{mailer};

    my $email = $c->stash->{ $self->stash_key };
    croak "Can't send email without a valid email structure"
      unless $email;

    # Default content type
    if ( $self->content_type and not $email->{content_type} ) {
        $email->{content_type} = $self->content_type;
    }

    my $header = $email->{header} || [];
    push @$header, ( 'To' => delete $email->{to} )
      if $email->{to};
    push @$header, ( 'Cc' => delete $email->{cc} )
      if $email->{cc};
    push @$header, ( 'From' => delete $email->{from} )
      if $email->{from};
    push @$header,
      ( 'Subject' => Encode::encode( 'MIME-Header', delete $email->{subject} ) )
      if $email->{subject};
    push @$header, ( 'Content-type' => $email->{content_type} )
      if $email->{content_type};

    my $parts = $email->{parts};
    my $body  = $email->{body};
    my $body_str = $email->{body_str};

    unless ( $parts or $body or $body_str ) {
        croak "Can't send email without parts or body, check stash";
    }

    my %mime = ( header => $header, attributes => {} );

    if ( $parts and ref $parts eq 'ARRAY' ) {
        $mime{parts} = $parts;
    }
    elsif ( defined $body ) {
        $mime{body} = $body;
    }
    else {
        $mime{body_str} = $body_str;
    }

    $mime{attributes}->{content_type} = $email->{content_type}
      if $email->{content_type};
    if (    $mime{attributes}
        and not $mime{attributes}->{charset}
        and $self->{default}->{charset} )
    {
        $mime{attributes}->{charset} = $self->{default}->{charset};
    }

    $mime{attributes}->{encoding} = $email->{encoding} 
        if $email->{encoding};

    my $message = $self->generate_message( $c, \%mime );

    if ($message) {
        my $return = sendmail( $message,
          {
            exists $email->{envelope_from} ? ( from => $email->{envelope_from} ) : (),
            exists $email->{envelope_to}   ? ( to   => $email->{envelope_to}   ) : (),
            transport => $self->_mailer_obj,
          } );

        # return is a Return::Value object, so this will stringify as the error
        # in the case of a failure.
        croak "$return" if !$return;
    }
    else {
        croak "Unable to create message";
    }
}

=item setup_attributes($c, $attr)

Merge attributes with the configured defaults. You can override this method to
return a structure to pass into L<generate_message> which subsequently
passes the return value of this method to Email::MIME->create under the
C<attributes> key.

=cut

sub setup_attributes {
    my ( $self, $c, $attrs ) = @_;

    my $default_content_type = $self->default->{content_type};
    my $default_charset      = $self->default->{charset}; 
    my $default_encoding     = $self->default->{encoding};

    my $e_m_attrs = {};

    if (   exists $attrs->{content_type}
        && defined $attrs->{content_type}
        && $attrs->{content_type} ne '' )
    {
        $c->log->debug( 'C::V::Email uses specified content_type '
              . $attrs->{content_type}
              . '.' )
          if $c->debug;
        $e_m_attrs->{content_type} = $attrs->{content_type};
    }
    elsif ( defined $default_content_type && $default_content_type ne '' ) {
        $c->log->debug(
            "C::V::Email uses default content_type $default_content_type.")
          if $c->debug;
        $e_m_attrs->{content_type} = $default_content_type;
    }

    if (   exists $attrs->{charset}
        && defined $attrs->{charset}
        && $attrs->{charset} ne '' )
    {
        $e_m_attrs->{charset} = $attrs->{charset};
    }
    elsif ( defined $default_charset && $default_charset ne '' ) {
        $e_m_attrs->{charset} = $default_charset;
    }

    if ( exists $attrs->{encoding}
         && defined $attrs->{encoding}
         && $attrs->{encoding} ne '' )
    {
        $c->log->debug(
        'C::V::Email uses specified encoding '
        . $attrs->{encoding}
        . '.' )
         if $c->debug;
         $e_m_attrs->{encoding} = $attrs->{encoding};
    }
     elsif ( defined $default_encoding && $default_encoding ne '' ) {
         $c->log->debug(
         "C::V::Email uses default encoding $default_encoding.")
         if $c->debug;
         $e_m_attrs->{encoding} = $default_encoding;
     }

    return $e_m_attrs;
}

=item generate_message($c, $attr)

Generate a message part, which should be an L<Email::MIME> object and return it.

Takes the attributes, merges with the defaults as necessary and returns a
message object.

=cut

sub generate_message {
    my ( $self, $c, $attr ) = @_;

    # setup the attributes (merge with defaults)
    $attr->{attributes} = $self->setup_attributes( $c, $attr->{attributes} );
    Email::MIME->create( %$attr );
}

=back


=head1 TROUBLESHOOTING

As with most things computer related, things break.  Email even more so.  
Typically any errors are going to come from using SMTP as your sending method,
which means that if you are having trouble the first place to look is at
L<Email::Sender::Transport::SMTP>.  This module is just a wrapper for L<Email::Sender::Simple>,
so if you get an error on sending, it is likely from there anyway.

If you are using SMTP and have troubles sending, whether it is authentication
or a very bland "Can't send" message, make sure that you have L<Net::SMTP> and,
if applicable, L<Net::SMTP::SSL> installed.

It is very simple to check that you can connect via L<Net::SMTP>, and if you
do have sending errors the first thing to do is to write a simple script
that attempts to connect.  If it works, it is probably something in your
configuration so double check there.  If it doesn't, well, keep modifying
the script and/or your mail server configuration until it does!

=head1 SEE ALSO

=head2 L<Catalyst::View::Email::Template> - Send fancy template emails with Cat

=head2 L<Catalyst::Manual> - The Catalyst Manual

=head2 L<Catalyst::Manual::Cookbook> - The Catalyst Cookbook

=head1 AUTHORS

J. Shirley <jshirley@gmail.com>

Alexander Hartmaier <abraxxa@cpan.org>

=head1 CONTRIBUTORS

(Thanks!)

Matt S Trout

Daniel Westermann-Clark

Simon Elliott <cpan@browsing.co.uk>

Roman Filippov

Lance Brown <lance@bearcircle.net>

Devin Austin <dhoss@cpan.org>

Chris Nehren <apeiron@cpan.org>

=head1 COPYRIGHT

Copyright (c) 2007 - 2009
the Catalyst::View::Email L</AUTHORS> and L</CONTRIBUTORS>
as listed above.

=head1 LICENSE

This library is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut

1;
